packages/product-json-to-csv/src/map-product-data.js

```diff
@@ -74,92 +106,257 @@ export default class ProductMapping {
+        if (!keyWord.suggestTokenizer) standard.push(keyWord.text)
+        else if (keyWord.suggestTokenizer.type === 'whitespace')
+          whitespace.push(keyWord.text.replace(/ /g, ' | '))
+      })
+      keywords[language] = [...standard, ...whitespace].join(';')
+    })
+    return keywords
+  }
+
+  _mapVariantProperties(variant: Variant, productType: ?ProductType): Object {
+    const mappedVariant: Object = _.reduce(
+      variant,
+      (acc: Object, value: any, property: string): Object => {
+        acc[property] = this._mapVariantProperty(property, value)
+        return acc
+      },
+      {}
+    )
+
+    // complete missing attributes
+    if (productType && productType.attributes)
+      productType.attributes.forEach((attribute: Object) => {
+        const mappedAttributes = mappedVariant.attributes || {}
+        if (_.isNil(mappedAttributes[attribute.name]))
+          mappedAttributes[attribute.name] = ''
+      })
+
+    return mappedVariant
+  }
+
+  _mapVariantProperty(property: string, value: any): any {
+    switch (property) {
+      case 'attributes':
+        return this._mapAttributes(value)
+      case 'prices':
+        return this._mapPrices(value)
+      case 'images':
+        return this._mapImages(value)
+      case 'availability':
+        return value.availableQuantity
+      case 'id':
+      case 'sku':
+      case 'key':
+        return value
+      default:
+        return undefined
+    }
+  }
+
+  _mapPrices(prices: Array<Price>): string {
+    return prices
+      .map((price: Price) => ProductMapping._mapPrice(price))
+      .join(this.multiValDel)
+  }
+
+  static _mapPrice(price: Price): string {
+    // Full price:
+    // 'country-currencyCode centAmount|discounted.centAmount customerGroup.name#channel.key$validFrom~validUntil'
+    let priceString: string = ''
+
+    if (price.country) priceString += `${price.country}-`
+
+    priceString += `${price.value.currencyCode} ${price.value.centAmount}`
+
+    if (price.discounted) priceString += `|${price.discounted.value.centAmount}`
+
+    if (price.customerGroup && price.customerGroup.name)
+      priceString += ` ${price.customerGroup.name}`
+
+    if (price.channel && price.channel.key)
+      priceString += `#${price.channel.key}`
+
+    if (price.validFrom) priceString += `$${price.validFrom}`
+
+    if (price.validUntil) priceString += `~${price.validUntil}`
+
+    return priceString
+  }
+
+  _mapLtextAttribute(name: string, value: Object): Object {
+    // create an object with value based on selected language
+    const mappedAttribute = {
+      [name]: value[this.lang],
+    }
+    // create object with translations indexed with keys like: attributeName.en
+    const labels = flatten({
+      [name]: value, // value contains object with localized labels
+    })
+
+    // add labels in format "attribName.en: labelEnValue" to all attributes
+    return _.merge(mappedAttribute, labels)
```

Comment:
I will use your preferred solution but from the performance point of view, cloning objects/arrays wherever you want to merge something is not the very efficient way how to do it. Lodash merge modifies object provided in the first parameter which can introduce side effects but if you know what are you doing it can speed up things a little bit

Reply:
Updated here 4e06beed22d16ca96402a4d8148c25e1b0d702cb